src/librustc/traits/select.rs

```diff
@@ -661,14 +665,15 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
         &mut self,
         stack: TraitObligationStackList<'o, 'tcx>,
         predicates: I,
+        recursion_depth: usize
```

Comment:
can we call this the `evaluation_depth` or something like that?

It also feels like there should be a comment that links to the explanation below of why we have a *distinct* recursion depth tracking in use here.